### PR TITLE
Karpenter Addon: adding AMI Selector, refactor to use NodeTemplate

### PR DESCRIPTION
**Pull request by youngjeong46**
*Issue #, if available:* #726

*Description of changes:*
- Adding AMI Selector for the Provisioner
- Using NodeTemplate in addition to the provisioner (needed to acommodate AMI Selector)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by elamaran11 — COMMENTED** [review]
LGTM. Could we also update the doc?

**Review by elamaran11 — DISMISSED** [review]
LGTM

**Comment by elamaran11** [comment]
@shapirov103 FYI

**Review by parkand1 — APPROVED** [review]
LGTM

**Review by shapirov103 — COMMENTED** [review]
A couple of tactical/code cleanliness comments.

File: lib/addons/karpenter/index.ts [hunk 1]
```diff
@@ -19,6 +19,40 @@ import { Cluster } from 'aws-cdk-lib/aws-eks';
  * Configuration options for the add-on
  */
 interface KarpenterAddOnProps extends HelmAddOnUserProps {
+    /**
+     * Taints for the provisioned nodes - Taints may prevent pods from scheduling if they are not tolerated by the pod.
+     */
+    taints?: {
```
[the remark this answers is not in the document]
Comment (youngjeong46) [reply lib/addons/karpenter/index.ts:25]: This didn't work - the provisioner CRD doesn't recognize TaintEffect interface and will cause an error.

File: lib/addons/karpenter/index.ts [hunk 2]
```diff
@@ -19,6 +19,40 @@ import { Cluster } from 'aws-cdk-lib/aws-eks';
...
+    startupTaints?: {
+        key: string,
+        value: string,
+        effect: "NoSchedule" | "PreferNoSchedule" | "NoExecute",
+    },
+
+    /**
+     * Labels applied to all nodes
+     */
+    labels?: {
```
[the remark this answers is not in the document]
Comment (youngjeong46) [reply lib/addons/karpenter/index.ts:45]: addressed.

File: lib/addons/karpenter/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (shapirov103) [reply lib/addons/karpenter/index.ts:25]: I proposed to only change the API layer for convenience. The backend will remain the same, you will need to convert TaintEffect to string. However, this functionality is already released, so my only comment here is: let's extracts the TaintType, put it in the spi types and reuse in the current form.

**Review by shapirov103 — COMMENTED** [review]
Looks good, please see my comment on taints.

**Comment by youngjeong46** [comment]
> Looks good, please see my comment on taints.

addressed!

**Review by shapirov103 — APPROVED** [review]
LGTM

**Pull request merged**